### ISSUE TITLE
ABEM naming method

### ISSUE THREAD

**Issue by Dan503** [bem-site/bem-method#668]
This is a BEM based naming convention I came up with that helps users that enjoy using Atomic Design principles.

````
[a/m/o]-blockName__elementName -modifierName
````
More details can be found in this CSS-Tricks article:
https://css-tricks.com/abem-useful-adaptation-bem/

**Comment by viT-1** [comment]
For **mixins support** you should have straight block__element with modifier accordance.
I prefer CSS 2.1 way founded on [AMCSS](https://amcss.github.io/):
`[data-blockName__elementName *= "-modifierForElementName-"]`
Hyphens in `-modifierName-` are needed to distinguish between variants, for example: `*= 'active'` and `*= 'inactive'`.
With attributes you can group multiple block__element modifiers without repeating long block__element names in markup.

For example: [tree__list](https://vit-1.github.io/iAMcss/samples/tree/list.xml)

**Comment by viT-1** [comment]
For nesting I prefer to use simple (**same weight**) selector such as:
`[data-blockName__elementName *= "^someParentBlockOrParentBlockStateModifier^"]`
You can see that I use different "parentheses" for modifier name to have accent on any context/nesting modifiers.

**Comment by Dan503** [comment]
Nah, I prefer the `.-modifier` selector. It is much simpler to write and using classes is also more efficient in browsers than using attribute selectors on everything.

**Comment by viT-1** [comment]
@Dan503 Can you show ABEM example with mixins (multiple blocks on one DOM element)?

**Comment by Dan503** [comment]
Is this what you mean? 😕

```html
<div class="a-one a-two -mod">
    <div class="a-one__elem a-two__elem -anotherMod"></div>
</div>
```

```scss
.a-one {
  &.-mod {}

  &__elem {
    .-mod > & {}
    &.-anotherMod {}
  }
}

.a-two {
  &.-mod {}

  &__elem {
    .-mod > & {}
    &.-anotherMod {}
  }
}

```

**Comment by viT-1** [comment]
> using classes is also more efficient in browsers than using attribute selectors
There will be a new version of browsers and the situation will change...

BTW, articles:
[selector strategy](https://gomakethings.com/javascript-selector-strategy/): 

> getElementById() can run about 15 million operations a second, compared to just 7 million per second for querySelector() in the latest version of Chrome. But that also means that querySelector() runs 7,000 operations a millisecond. A millisecond. 

[benchmark](https://jsperf.com/queryselectorall-vs-getelementbyid/32)

**Comment by Dan503** [comment]
There is a GIGANTIC difference between selecting an element by it's ID and selecting it by a random attribute you made up.

- `#selector` is fastest but is a nightmare to maintain.
- `.selector` is a bit slower but is far more maintainable.
- `[selector]` is by far the slowest performing way to write a selector. I only use it if I have to.

**Comment by viT-1** [comment]
We have js-frameworks for "quering" DOM (components binding way), therefore we can use data-attributes for front representation (styling).

**Comment by Dan503** [comment]
Actually yeah, I may need to take that back. I tried running the benchmark test and the data attribute was pretty comparable to classes surprisingly.

So taking performance out of the question,  I still think the classes are a nicer syntax to write.

**Comment by viT-1** [comment]
About your SASS example:
You think on global modifiers. For me it is something like a global namespace in js.
If i will use a-two__elem in -anotherMod state I will recive a bug about wrong state of a-one__elem

**Comment by viT-1** [comment]
Pretty article from @toddmotto: [Stop toggling classes with JS, use behaviour driven DOM manipulation with data-states](https://ultimatecourses.com/blog/stop-toggling-classes-with-js-use-behaviour-driven-dom-manipulation-with-data-states) Published on 17 Dec 2013!

**Comment by veged** [comment]
Thanks @Dan503 for taking the time to write this up — and sorry for the very late reply.

We're honestly not big fans of Atomic Design as a foundation underneath BEM, and that's the main reason we don't want to bake `a-`/`m-`/`o-` prefixes (or any equivalent) into the methodology itself.

The core concern is that Atomic Design hard-codes a three-level hierarchy — atom → molecule → organism — into the *name* of every entity. In BEM the block tree is intentionally open-ended: any block can contain any other block at any depth, and a block's "rank" emerges from how it's used on a given project, not from its identifier.

This bites in practice when projects evolve:

- A block that started as an "atom" often grows to wrap a few smaller things and effectively becomes a "molecule" — and the other way around, parts get extracted and demoted. With ABEM-style prefixes every such refactoring turns into a global rename across HTML, CSS, JS, templates and tooling, instead of staying local.
- Two projects can reasonably classify the "same" block differently (a `button` may be an atom in one project but a molecule in another the moment it embeds an icon block). Encoding that classification in the name turns a portable block into a project-specific one.
- The BEM tree is meant to grow naturally — adding a new layer of nesting shouldn't require inventing a new prefix tier above "organism".

We'd rather keep block names neutral (`button`, `search-form`, `header`) and let composition be expressed by the structure of the block tree, the file structure and dependencies, without locking it into the naming scheme.

That said, BEM doesn't *forbid* combining it with Atomic ideas at the project level: if a team finds value in `a-`/`m-`/`o-` prefixes for their own workflow, that's totally their call — the [naming convention](https://bem.info/ru/methodology/naming-convention/) explicitly allows custom schemes via `bem-naming`. We just don't want to adopt it into the methodology itself.

Closing as won't-fix — thanks again for the proposal.
